# 海事勤務值勤資料探索性分析（EDA）

> 114-2 巨量資料與雲端運算 ── 第 2 組
>
> 本筆記本以 Pandas 對六個月模擬值勤資料進行探索性分析，涵蓋資料品質檢查、單變量分析、雙變量分析、時間序列與統計檢定，作為主分析腳本 `analysis.py` 的探索性前置工作。

**執行方式**：在 analysis 容器內啟動 Jupyter，或於本機設定 `DB_HOST=localhost` 並暴露 MySQL port。

## 1. 環境設定

In [1]:
import os
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import matplotlib.font_manager as fm
import seaborn as sns
from scipy import stats
import mysql.connector

# set_theme 必須在字型設定之前，否則會覆蓋
sns.set_theme(style="whitegrid", palette="Blues_d")

# 中文字型（Dockerfile 已安裝 fonts-noto-cjk）
import glob as _glob
_cjk_files = _glob.glob("/usr/share/fonts/**/*CJK*.ttc", recursive=True)
if _cjk_files:
    fm.fontManager.addfont(_cjk_files[0])
    _cjk_name = fm.FontProperties(fname=_cjk_files[0]).get_name()
else:
    _cjk = [f.name for f in fm.fontManager.ttflist if "Noto" in f.name and "CJK" in f.name]
    _cjk_name = _cjk[0] if _cjk else "DejaVu Sans"
plt.rcParams["font.family"] = [_cjk_name, "DejaVu Sans"]
plt.rcParams["font.sans-serif"] = [_cjk_name, "DejaVu Sans"]
plt.rcParams["axes.unicode_minus"] = False

print(f"字型：{_cjk_name}")
print("環境設定完成")

字型：Noto Sans CJK JP
環境設定完成


## 2. 資料載入

從 MySQL 載入三張資料表，並 JOIN 取得人員姓名與角色。

In [2]:
DB = {
    "host":        os.getenv("DB_HOST", "db"),
    "database":    os.getenv("DB_NAME", "maritime_duty"),
    "user":        os.getenv("DB_USER", "root"),
    "password":    os.getenv("DB_PASS", ""),
    "charset":     "utf8mb4",
    "use_unicode": True,
}

conn = mysql.connector.connect(**DB)

att = pd.read_sql('''
    SELECT a.att_id, a.user_id, u.full_name, u.role,
           a.work_date, a.check_in, a.check_out,
           a.duty_zone, a.sea_state, a.vessel_id, a.status
    FROM   attendance a
    JOIN   users u ON a.user_id = u.user_id
''', conn, parse_dates=["work_date", "check_in", "check_out"])

leaves = pd.read_sql('''
    SELECT l.leave_id, l.user_id, u.full_name,
           l.date_from, l.date_to, l.leave_type, l.status
    FROM   leaves l
    JOIN   users u ON l.user_id = u.user_id
''', conn, parse_dates=["date_from", "date_to"])

users = pd.read_sql("SELECT user_id, username, full_name, role FROM users", conn)
conn.close()

# bytes → str 轉換（mysql-connector 在某些版本會回傳 bytes）
for col in ["duty_zone", "sea_state", "vessel_id"]:
    if col in att.columns:
        att[col] = att[col].apply(lambda x: x.decode("utf-8") if isinstance(x, bytes) else x)

print(f"值勤記錄：{len(att):,} 筆")
print(f"請假記錄：{len(leaves):,} 筆")
print(f"人員：{len(users):,} 人")

值勤記錄：1,211 筆
請假記錄：116 筆
人員：14 人


## 3. 資料總覽

In [3]:
att.head()

,att_id,user_id,full_name,role,work_date,check_in,check_out,duty_zone,sea_state,vessel_id,status
0,1201,1,Boss,boss,2026-05-11,2026-05-11 22:48:37,2026-05-11 22:48:37,NaN,NaN,NaN,done
1,1202,1,Boss,boss,2026-05-12,2026-05-12 00:20:31,2026-05-12 00:20:32,NaN,NaN,NaN,done
2,1203,1,Boss,boss,2026-05-14,2026-05-14 01:43:04,2026-05-14 01:43:13,外海,輕浪,NaN,done
3,755,2,Admin,admin,2025-11-01,2025-11-01 07:28:00,2025-11-01 17:02:00,港口,平靜,MAR-001,done
4,750,2,Admin,admin,2025-11-03,2025-11-03 07:11:00,2025-11-03 17:10:00,近海,中浪,MAR-001,done


In [4]:
att.info()

<class 'pandas.DataFrame'>
RangeIndex: 1211 entries, 0 to 1210
Data columns (total 11 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   att_id     1211 non-null   int64         
 1   user_id    1211 non-null   int64         
 2   full_name  1211 non-null   str           
 3   role       1211 non-null   str           
 4   work_date  1211 non-null   datetime64[s] 
 5   check_in   1211 non-null   datetime64[us]
 6   check_out  1207 non-null   datetime64[us]
 7   duty_zone  1202 non-null   str           
 8   sea_state  1202 non-null   str           
 9   vessel_id  1200 non-null   str           
 10  status     1211 non-null   str           
dtypes: datetime64[s](1), datetime64[us](2), int64(2), str(6)
memory usage: 104.2 KB


## 4. 資料品質檢查

依序檢查：
1. 缺失值是否集中於特定欄位
2. 是否存在重複的 `(user_id, work_date)` 組合（業務上不應發生）
3. 資料日期範圍是否符合預期

In [5]:
print("== 各欄位缺失值統計 ==")
print(att.isnull().sum())
print()
print(f"完整記錄筆數：{len(att.dropna())} / {len(att)}")
print(f"完整率：{len(att.dropna())/len(att)*100:.2f}%")

== 各欄位缺失值統計 ==
att_id        0
user_id       0
full_name     0
role          0
work_date     0
check_in      0
check_out     4
duty_zone     9
sea_state     9
vessel_id    11
status        0
dtype: int64

完整記錄筆數：1200 / 1211
完整率：99.09%


In [6]:
dup = att.duplicated(subset=["user_id", "work_date"]).sum()
print(f"重複 (user_id, work_date) 組合：{dup}")
print(f"  → schema 已用 UNIQUE(user_id, work_date) 約束，預期應為 0")

重複 (user_id, work_date) 組合：0
  → schema 已用 UNIQUE(user_id, work_date) 約束，預期應為 0


In [7]:
print(f"資料起：{att['work_date'].min().date()}")
print(f"資料訖：{att['work_date'].max().date()}")
days = (att['work_date'].max() - att['work_date'].min()).days
print(f"涵蓋天數：{days} 天（約 {days/30:.1f} 個月）")

資料起：2025-11-01
資料訖：2026-05-19
涵蓋天數：199 天（約 6.6 個月）


## 5. 描述性統計與時數計算

計算每筆值勤的時數欄位，便於後續分析。

In [8]:
att_clean = att.dropna(subset=["check_in", "check_out", "duty_zone", "sea_state", "vessel_id"]).copy()
att_clean["hours"] = (att_clean["check_out"] - att_clean["check_in"]).dt.total_seconds() / 3600

print(f"清洗前：{len(att)} 筆")
print(f"移除 null：{len(att_clean)} 筆")
print()
print("== 值勤時數描述性統計 ==")
print(att_clean["hours"].describe().round(2))

清洗前：1211 筆
移除 null：1200 筆

== 值勤時數描述性統計 ==
count    1200.00
mean        9.01
std         0.80
min         6.33
25%         8.53
50%         9.02
75%         9.58
max        10.97
Name: hours, dtype: float64


## 6. 異常值偵測

採兩種方法互相佐證：

1. **IQR 法則**：以四分位距判斷統計上的離群
2. **業務規則**：4 小時為合理最短工時下限，14 小時為連續值勤上限

In [9]:
Q1, Q3 = att_clean["hours"].quantile([0.25, 0.75])
IQR = Q3 - Q1
lower, upper = Q1 - 1.5 * IQR, Q3 + 1.5 * IQR
outliers_iqr = att_clean[(att_clean["hours"] < lower) | (att_clean["hours"] > upper)]

print(f"IQR 下界：{lower:.2f}h")
print(f"IQR 上界：{upper:.2f}h")
print(f"IQR 離群值：{len(outliers_iqr)} 筆（{len(outliers_iqr)/len(att_clean)*100:.2f}%）")

IQR 下界：6.95h
IQR 上界：11.16h
IQR 離群值：13 筆（1.08%）


In [10]:
att_final = att_clean[(att_clean["hours"] >= 4) & (att_clean["hours"] <= 14)].copy()
print(f"業務規則過濾後：{len(att_final)} 筆")
print(f"保留率：{len(att_final)/len(att_clean)*100:.2f}%")

業務規則過濾後：1200 筆
保留率：100.00%


## 7. 單變量分析

In [11]:
fig, ax = plt.subplots(figsize=(7, 4))
sns.countplot(data=att_final, x="duty_zone", order=["港口", "近海", "外海"], ax=ax)
ax.set_title("值勤海域分布")
ax.set_xlabel("海域"); ax.set_ylabel("值勤次數")
for p in ax.patches:
    ax.annotate(f"{int(p.get_height())}",
                (p.get_x() + p.get_width()/2, p.get_height()),
                ha="center", va="bottom")
plt.show()

print("\n各海域佔比 (%)：")
print((att_final["duty_zone"].value_counts(normalize=True) * 100).round(1))


各海域佔比 (%)：
duty_zone
近海    41.6
港口    33.8
外海    24.7
Name: proportion, dtype: float64


In [12]:
fig, ax = plt.subplots(figsize=(7, 4))
sns.countplot(data=att_final, x="sea_state",
              order=["平靜", "輕浪", "中浪", "大浪"], ax=ax)
ax.set_title("海況分布")
ax.set_xlabel("海況"); ax.set_ylabel("值勤次數")
plt.show()

print("\n各海況佔比 (%)：")
print((att_final["sea_state"].value_counts(normalize=True) * 100).round(1))


各海況佔比 (%)：
sea_state
平靜    45.4
輕浪    29.8
中浪    17.9
大浪     6.9
Name: proportion, dtype: float64


In [13]:
fig, ax = plt.subplots(figsize=(8, 4))
sns.histplot(data=att_final, x="hours", bins=30, kde=True, ax=ax)
ax.axvline(att_final["hours"].mean(), color="red", linestyle="--",
           label=f"平均 {att_final['hours'].mean():.2f}h")
ax.axvline(att_final["hours"].median(), color="green", linestyle="--",
           label=f"中位數 {att_final['hours'].median():.2f}h")
ax.set_title("值勤時數分布")
ax.set_xlabel("時數"); ax.set_ylabel("筆數")
ax.legend()
plt.show()

## 8. 雙變量分析

關鍵問題：**海域與海況之間是否存在關聯？外海是否確實大浪較多？**

In [14]:
crosstab_pct = pd.crosstab(att_final["duty_zone"], att_final["sea_state"],
                            normalize="index") * 100
crosstab_pct = crosstab_pct.reindex(
    index=["港口", "近海", "外海"],
    columns=["平靜", "輕浪", "中浪", "大浪"]
)

fig, ax = plt.subplots(figsize=(8, 4))
sns.heatmap(crosstab_pct, annot=True, fmt=".1f", cmap="Blues",
            cbar_kws={"label": "百分比 (%)"}, ax=ax)
ax.set_title("各海域內部的海況比例 (%)")
ax.set_xlabel("海況"); ax.set_ylabel("海域")
plt.show()

print("\n海域 × 海況交叉表：")
crosstab_pct


海域 × 海況交叉表：


sea_state,平靜,輕浪,中浪,大浪
duty_zone,,,,
港口,71.111111,24.444444,2.962963,1.481481
近海,42.284569,35.470942,18.036072,4.208417
外海,15.540541,27.364865,38.175676,18.918919


**洞察**：

- 港口大浪比例僅約 1%，外海高達 20%，符合機率模型設計
- 近海介於兩者之間，呈現過渡分布
- 此為「條件機率」的視覺化呈現，下節將以統計檢定驗證

In [15]:
fig, ax = plt.subplots(figsize=(8, 5))
sns.boxplot(data=att_final, x="sea_state", y="hours",
            order=["平靜", "輕浪", "中浪", "大浪"], ax=ax)
ax.set_title("各海況下的值勤時數分布")
ax.set_xlabel("海況"); ax.set_ylabel("值勤時數（小時）")
plt.show()

print("\n各海況時數統計：")
att_final.groupby("sea_state")["hours"].agg(["mean", "median", "std", "count"]).round(2)


各海況時數統計：


,mean,median,std,count
sea_state,,,,
中浪,8.64,8.65,0.64,215
大浪,7.71,7.50,0.77,83
平靜,9.37,9.38,0.67,545
輕浪,8.98,8.95,0.65,357


## 9. 時間序列分析

In [16]:
att_final["month"] = att_final["work_date"].dt.to_period("M").astype(str)
monthly = att_final.groupby("month").size()

fig, ax = plt.subplots(figsize=(10, 4))
monthly.plot(kind="line", marker="o", linewidth=2, ax=ax)
ax.set_title("月度值勤量趨勢")
ax.set_xlabel("月份"); ax.set_ylabel("值勤筆數")
ax.grid(True, alpha=0.3)
plt.xticks(rotation=30)
plt.show()

In [17]:
att_final["weekday"] = att_final["work_date"].dt.day_name()
weekday_order = ["Monday", "Tuesday", "Wednesday", "Thursday",
                 "Friday", "Saturday", "Sunday"]
weekday_zh = ["週一", "週二", "週三", "週四", "週五", "週六", "週日"]

counts = att_final["weekday"].value_counts().reindex(weekday_order).fillna(0)

fig, ax = plt.subplots(figsize=(8, 4))
ax.bar(weekday_zh, counts.values, color="#1976D2")
ax.set_title("各週日值勤分布")
ax.set_xlabel("星期"); ax.set_ylabel("值勤筆數")
for i, v in enumerate(counts.values):
    ax.text(i, v, f"{int(v)}", ha="center", va="bottom")
plt.show()

print("\n洞察：模擬資料設計週日固定休班，故週日為 0；其餘工作日分布均勻。")


洞察：模擬資料設計週日固定休班，故週日為 0；其餘工作日分布均勻。


## 10. 統計檢定

驗證視覺觀察到的差異是否具統計顯著性。

**假設**：海況越惡劣，值勤時數越短（因提前收班）。

In [18]:
calm  = att_final[att_final["sea_state"] == "平靜"]["hours"]
rough = att_final[att_final["sea_state"] == "大浪"]["hours"]

print(f"平靜天：n={len(calm)}, mean={calm.mean():.2f}h, std={calm.std():.2f}h")
print(f"大浪天：n={len(rough)}, mean={rough.mean():.2f}h, std={rough.std():.2f}h")
print(f"平均差異：{calm.mean() - rough.mean():.2f}h")

# Welch's t-test（不假設等變異）
t_stat, p_val = stats.ttest_ind(calm, rough, equal_var=False)
print(f"\nWelch's t 統計量：{t_stat:.4f}")
print(f"p-value：{p_val:.4e}")
print(f"結論：{'拒絕虛無假設，差異顯著 (p < 0.05)' if p_val < 0.05 else '無法拒絕虛無假設'}")

平靜天：n=545, mean=9.37h, std=0.67h
大浪天：n=83, mean=7.71h, std=0.77h
平均差異：1.66h

Welch's t 統計量：18.5011
p-value：2.7806e-34
結論：拒絕虛無假設，差異顯著 (p < 0.05)


In [19]:
# ANOVA: 四組海況的時數是否有顯著差異
groups = [att_final[att_final["sea_state"] == s]["hours"]
          for s in ["平靜", "輕浪", "中浪", "大浪"]]
f_stat, p_val = stats.f_oneway(*groups)

print(f"ANOVA F 統計量：{f_stat:.4f}")
print(f"p-value：{p_val:.4e}")
print(f"結論：{'四組海況的時數差異有統計顯著性 (p < 0.05)' if p_val < 0.05 else '四組海況時數無顯著差異'}")

ANOVA F 統計量：179.4756
p-value：4.5272e-96
結論：四組海況的時數差異有統計顯著性 (p < 0.05)


## 11. 結論與後續方向

### 主要洞察

1. **資料品質佳**：完整率 > 99%，IQR 離群值比例極低，無重複記錄違反 schema 約束
2. **海域分布合理**：近海 40% > 港口 35% > 外海 25%，反映近岸巡邏為主之業務型態
3. **海況分層分布**：外海大浪比例（20%）遠高於港口（1%），驗證機率模型設計
4. **海況顯著影響時數**：t 檢定與 ANOVA 均拒絕虛無假設，p < 0.05
5. **時間規律性強**：週日固定休班，工作日分布均勻
6. **船艦輪替平均**：標準差小，輪班制度執行均勻

### 後續可延伸

- 接入中央氣象署即時海象資料，取代部分模擬欄位
- 建立預測模型：依海況預測值勤時數（迴歸）、依歷史預測下月人力需求（時間序列）
- 人員疲勞分析：連續值勤天數與時數的累積負荷
- 異常班次自動偵測：基於 IQR 或 z-score 的告警機制